### ISSUE TITLE
💬 RFC:  Tracking group ownership in Backstage Catalog

### ISSUE THREAD

**Issue by alde** [backstage/backstage#18823]
### 🔖 Need

Currently in the Catalog users in a group can only be members. There is no distinction between someone with basic access and someone with elevated privileges (such as adding members to the group). This means that there is a loss of metadata from sources such as LDAP, where you can have multiple owners and members of groups, or GitHub where you can have different access levels.

Because there can be several owners of an LDAP group, the existing `spec.owner` field will probably not be enough, and it could be confusing to deviate from this being a single `entityRef`.

In LDAP the owner field is a predefined optional field provided by the `groupOfNames` objectClass, which also provides the member field which we depend on.

### hasOwner/isOwner type relations
Add a new relation, with `hasOwner`/`isOwner` type. This would result in a lot more relations with many user relations appearing twice (once with `type: hasMember`, and once with `type: hasOwner`), but would require no schema modifications.

This relationship could be expressed for other entities too, not just groups. And could be useful for querying through the catalog using the relations.

### Metadata on Relationships
We could keep the existing `hasMember`/`memberOf` relations but extend the relationship type with additional metadata. This has been something that has been requested in other places to indicate roles around the company before, and using each company's individual terminology. This could be a good start to be able to indicate users being `memberOf` as a different role, however some thought is needed here about if having two of the same relationship type is possible in the current setup

`{ type: ‘memberOf’, metadata: { role: ‘owner’ }, targetEntityRef: ‘user:default/dybeck’ }`

### 👀 Have you spent some time to check if this RFC has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

**Comment by freben** [comment]
Hi!

No matter what the input model is (as in, what the shape of the `spec` turns out to be), it'd still make sense to reuse the existing `ownedBy`/`ownerOf` relation type pair as the output right? Unless the intent is to treat this ownership as a distinctly different form of ownership than eg component ownership.

If it really is a philosophically different form of ownership it can make sense to choose another relation type pair of names. But then they should probably be more distinctly different to reduce risk of confusion.

**Comment by alde** [comment]
Yes, you're right it should be ownedBy/ownerOf

**Comment by sennyeya** [comment]
Thanks for this! 

Couple of thoughts here, 

mharrah had mentioned this in [this comment](https://github.com/backstage/backstage/issues/17996#issuecomment-1626338758), but group "ownership" could be a contentious naming that we may want to stay away from as well. I think if we go with `ownerOf`/`ownedBy`, it should be consistent with the rest of ownership which would also be single-user based. 

To avoid that multi-ownership, my vote would be for the alternative approach with metadata on the relations. That would allow for different access levels, primary/secondary owners, administrators, etc. For the case of users not in the group being LDAP owners, I'd propose a new `hasAccess` attribute to store all access-level (`read`/`write`) data.

The way I see it, our goal isn't to model what's in upstream systems, it's to simplify that data into a comprehensible base format that everyone can _generally_ agree about, a new relation like `hasAccess` could standardize that into `read`/`write`/`admin` instead of getting caught up in company by company wording and definitions. Yes, it would simplify models from complex systems like Github (where would maintainer vs admin vs write sit?), but I don't think our goal should be to model all of that data, but a useful subset. 

Of course, this opens concerns about mapping X complex dimensions into Y simpler dimensions and creating a new standard (https://xkcd.com/927/), but I think it would be a good direction towards capturing a base set of important information instead of continuing to fracture it.

This might also conflict with thinking about RBAC or FGAC, but I think that's also a good conversation to be had as we're talking about admin access on groups.

**Comment by Rugvip** [comment]
I wanna throw in a +1 for using an additional relation for this, `ownedBy`/`ownerOf` to be precise. We did leave room for relation metadata in the data model, but it's still a scary concept that I think we should hold off adding until it's really needed. It adds significant complexity, especially considering how you'd evolve the relation metadata schema.

In this particular case I think it's also useful to be able to model a group with a set of members but with an owner that is not necessarily member of that group. What plugins would typically do with the knowledge of who is the owner of a group I feel would be defaults for access control (permission plugin) rules, as well as a point of contact for someone wanting to make changes to a group. For both of those cases I don't think the owner necessarily must be a member of the group itself.

**Comment by alde** [comment]
Internally we modified the GroupTransformer and the Group spec to include ownership, and then a processor to create a ownedBy/ownerOf relationship on both users and groups.

This works well for cases where someone is an owner but not a member, but does create additional data when people are both.

**Comment by pjungermann** [comment]
I've implemented the leadership aspect for our org which probably has an overlap, maybe not 100% for what you had in mind with ownership, though.

```
Group:
spec.leaders (array)

User:
spec.leaderOf (array)

relations:
ledBy / leaderOf
```

An alternative based on lead vs leader (some would have preferred lead):

```
Group:
spec.leads (array)

User:
spec.leadOf (array)

relations:
ledBy / leadOf
```

Still open for adjustments.

While doing so, I found it not so easy to extend an existing kind as you cannot reference it in the schema. Only Entity and a few others have been added to ajv for validation and URL refs are not supported either. I will create a separate issue for that topic, though.
